force-app/main/default/classes/ContactAccountHandlerTest.cls

```diff
@@ -0,0 +1,53 @@
+@isTest
+public with sharing class ContactAccountHandlerTest {
+    @isTest static void isPrimaryTest(){
+        Account acc1 = new Account(Name='Demo');
+        insert acc1;
+        
+        Account acc2 = new Account(Name='Demo 1', ParentId=acc1.Id);
+        insert acc2;
+        
+        List<Contact> contactList = new List<Contact>();
+        for(Integer i=0; i<10; i++){
+            Contact contact = new Contact();
+            contact.LastName='Demo Contact '+i;
+            contact.AccountId=acc1.id;
+            if(i>5){
+                contact.Is_Primary__c= true;
+            }
+            contactList.add(contact);
+        }
+        
+        Test.startTest();
+        List<Database.SaveResult> results = Database.insert(contactList, false);
+        Test.stopTest();
+        for (Database.SaveResult sr : results) {
+            if (!sr.isSuccess()) {
+                for(Database.Error err : sr.getErrors()) {                
+                    System.assertEquals(true,err.getMessage().contains('Primary Contact Already Exists'));
```

Comment:
use counter variable to calculate faild contacts numbers and perform assertion on counter variable,
according to this code there are 4 contacts must be failed at the time of insertion